In [112]:
import pandas as pd
import numpy as np 
import matplotlib.pyplot as plt
import seaborn as sns




In [113]:
df = pd.read_csv('house_prices.csv')

In [114]:
df


area_type   availability                  location  \
0      Super built-up  Area         19-Dec  Electronic City Phase II   
1                Plot  Area  Ready To Move          Chikka Tirupathi   
2            Built-up  Area  Ready To Move               Uttarahalli   
3      Super built-up  Area  Ready To Move        Lingadheeranahalli   
4      Super built-up  Area  Ready To Move                  Kothanur   
...                     ...            ...                       ...   
13315        Built-up  Area  Ready To Move                Whitefield   
13316  Super built-up  Area  Ready To Move             Richards Town   
13317        Built-up  Area  Ready To Move     Raja Rajeshwari Nagar   
13318  Super built-up  Area         18-Jun           Padmanabhanagar   
13319  Super built-up  Area  Ready To Move              Doddathoguru   

            size  society total_sqft  bath  balcony   price  
0          2 BHK  Coomee        1056   2.0      1.0   39.07  
1      4 Bedroom  Theanmp       2600   5.0      3.0  120.00  
2          3 BHK      NaN       1440   2.0      3.0   62.00  
3          3 BHK  Soiewre       1521   3.0      1.0   95.00  
4          2 BHK      NaN       1200   2.0      1.0   51.00  
...          ...      ...        ...   ...      ...     ...  
13315  5 Bedroom  ArsiaEx       3453   4.0      0.0  231.00  
13316      4 BHK      NaN       3600   5.0      NaN  400.00  
13317      2 BHK  Mahla T       1141   2.0      1.0   60.00  
13318      4 BHK  SollyCl       4689   4.0      1.0  488.00  
13319      1 BHK      NaN        550   1.0      1.0   17.00  

[13320 rows x 9 columns]

In [115]:
df2  = df.drop(['area_type','society','availability' ,'balcony'] , axis=1)

In [116]:
df2

location       size total_sqft  bath   price
0      Electronic City Phase II      2 BHK       1056   2.0   39.07
1              Chikka Tirupathi  4 Bedroom       2600   5.0  120.00
2                   Uttarahalli      3 BHK       1440   2.0   62.00
3            Lingadheeranahalli      3 BHK       1521   3.0   95.00
4                      Kothanur      2 BHK       1200   2.0   51.00
...                         ...        ...        ...   ...     ...
13315                Whitefield  5 Bedroom       3453   4.0  231.00
13316             Richards Town      4 BHK       3600   5.0  400.00
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0   60.00
13318           Padmanabhanagar      4 BHK       4689   4.0  488.00
13319              Doddathoguru      1 BHK        550   1.0   17.00

[13320 rows x 5 columns]

In [117]:
df2.isnull().sum()

location       1
size          16
total_sqft     0
bath          73
price          0
dtype: int64

In [118]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 13320 entries, 0 to 13319
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13319 non-null  object 
 1   size        13304 non-null  object 
 2   total_sqft  13320 non-null  object 
 3   bath        13247 non-null  float64
 4   price       13320 non-null  float64
dtypes: float64(2), object(3)
memory usage: 520.4+ KB


out off 13320 entries only few are null so can drop them

In [119]:
df3 = df2.dropna()

In [120]:
df3.isnull().sum()

location      0
size          0
total_sqft    0
bath          0
price         0
dtype: int64

In [121]:
df3 = df3.copy()
# used this cause df3 is just a view , for working on a dataframe need to make a dataframe by using .copy()
df3['bhk'] = df3['size'].apply(lambda x : x.split(' ')[0])

In [122]:
df3

location       size total_sqft  bath   price bhk
0      Electronic City Phase II      2 BHK       1056   2.0   39.07   2
1              Chikka Tirupathi  4 Bedroom       2600   5.0  120.00   4
2                   Uttarahalli      3 BHK       1440   2.0   62.00   3
3            Lingadheeranahalli      3 BHK       1521   3.0   95.00   3
4                      Kothanur      2 BHK       1200   2.0   51.00   2
...                         ...        ...        ...   ...     ...  ..
13315                Whitefield  5 Bedroom       3453   4.0  231.00   5
13316             Richards Town      4 BHK       3600   5.0  400.00   4
13317     Raja Rajeshwari Nagar      2 BHK       1141   2.0   60.00   2
13318           Padmanabhanagar      4 BHK       4689   4.0  488.00   4
13319              Doddathoguru      1 BHK        550   1.0   17.00   1

[13246 rows x 6 columns]

In [123]:
df3['bhk'].unique()

array(['2', '4', '3', '6', '1', '8', '7', '5', '11', '9', '27', '10',
       '19', '16', '43', '14', '12', '13', '18'], dtype=object)

In [124]:
df3['total_sqft'].unique()

array(['1056', '2600', '1440', ..., '1133 - 1384', '774', '4689'],
      dtype=object)

In [125]:
def is_float(x):
    try:
        float(x)
    except:
        return False
    return True

In [126]:
wrong_format_total_sqft = df3[~df3['total_sqft'].apply(lambda x : is_float(x))]

In [127]:
wrong_format_total_sqft

location   size   total_sqft  bath    price bhk
30              Yelahanka  4 BHK  2100 - 2850   4.0  186.000   4
122                Hebbal  4 BHK  3067 - 8156   4.0  477.000   4
137    8th Phase JP Nagar  2 BHK  1042 - 1105   2.0   54.005   2
165              Sarjapur  2 BHK  1145 - 1340   2.0   43.490   2
188              KR Puram  2 BHK  1015 - 1540   2.0   56.800   2
...                   ...    ...          ...   ...      ...  ..
12975          Whitefield  2 BHK   850 - 1060   2.0   38.190   2
12990      Talaghattapura  3 BHK  1804 - 2273   3.0  122.000   3
13059              Harlur  2 BHK  1200 - 1470   2.0   72.760   2
13265               Hoodi  2 BHK  1133 - 1384   2.0   59.135   2
13299          Whitefield  4 BHK  2830 - 2882   5.0  154.500   4

[190 rows x 6 columns]

In [128]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13246 non-null  object 
 1   size        13246 non-null  object 
 2   total_sqft  13246 non-null  object 
 3   bath        13246 non-null  float64
 4   price       13246 non-null  float64
 5   bhk         13246 non-null  object 
dtypes: float64(2), object(4)
memory usage: 724.4+ KB


In [129]:
def convert_total_sqft_data(x):
    try:
     return (int(x.split('-')[0]) + int(x.split('-')[1]))/2
    except:
       return x

 

In [130]:
wrong_format_total_sqft['total_sqft'] = wrong_format_total_sqft['total_sqft'].apply(lambda x : convert_total_sqft_data(x))

C:\Users\SWARAJ\AppData\Local\Temp\ipykernel_9412\3926602575.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  wrong_format_total_sqft['total_sqft'] = wrong_format_total_sqft['total_sqft'].apply(lambda x : convert_total_sqft_data(x))


In [131]:
wrong_format_total_sqft

location   size total_sqft  bath    price bhk
30              Yelahanka  4 BHK     2475.0   4.0  186.000   4
122                Hebbal  4 BHK     5611.5   4.0  477.000   4
137    8th Phase JP Nagar  2 BHK     1073.5   2.0   54.005   2
165              Sarjapur  2 BHK     1242.5   2.0   43.490   2
188              KR Puram  2 BHK     1277.5   2.0   56.800   2
...                   ...    ...        ...   ...      ...  ..
12975          Whitefield  2 BHK      955.0   2.0   38.190   2
12990      Talaghattapura  3 BHK     2038.5   3.0  122.000   3
13059              Harlur  2 BHK     1335.0   2.0   72.760   2
13265               Hoodi  2 BHK     1258.5   2.0   59.135   2
13299          Whitefield  4 BHK     2856.0   5.0  154.500   4

[190 rows x 6 columns]

In [132]:
df3['total_sqft'] = df3['total_sqft'].apply(lambda x : convert_total_sqft_data(x))

In [133]:
df3.loc[30]
# location 30

location      Yelahanka
size              4 BHK
total_sqft       2475.0
bath                4.0
price             186.0
bhk                   4
Name: 30, dtype: object

In [134]:
def sqmeter_to_num(x):
    try:
        return int(float(x.split('S')[0]))*10.7639
    except:
        return x

In [135]:
df3['total_sqft'] = df3['total_sqft'].apply(lambda x : sqmeter_to_num(x))

In [136]:
df3.loc[410]

location       Kengeri
size             1 BHK
total_sqft    365.9726
bath               1.0
price             18.5
bhk                  1
Name: 410, dtype: object

In [137]:
df4 = df3.copy()

In [138]:
df4

location       size  total_sqft  bath   price bhk
0      Electronic City Phase II      2 BHK  11366.6784   2.0   39.07   2
1              Chikka Tirupathi  4 Bedroom    27986.14   5.0  120.00   4
2                   Uttarahalli      3 BHK   15500.016   2.0   62.00   3
3            Lingadheeranahalli      3 BHK  16371.8919   3.0   95.00   3
4                      Kothanur      2 BHK    12916.68   2.0   51.00   2
...                         ...        ...         ...   ...     ...  ..
13315                Whitefield  5 Bedroom  37167.7467   4.0  231.00   5
13316             Richards Town      4 BHK    38750.04   5.0  400.00   4
13317     Raja Rajeshwari Nagar      2 BHK  12281.6099   2.0   60.00   2
13318           Padmanabhanagar      4 BHK  50471.9271   4.0  488.00   4
13319              Doddathoguru      1 BHK    5920.145   1.0   17.00   1

[13246 rows x 6 columns]

In [139]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 6 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   location    13246 non-null  object 
 1   size        13246 non-null  object 
 2   total_sqft  13246 non-null  object 
 3   bath        13246 non-null  float64
 4   price       13246 non-null  float64
 5   bhk         13246 non-null  object 
dtypes: float64(2), object(4)
memory usage: 1.2+ MB


In [140]:
df4.loc[410]

location       Kengeri
size             1 BHK
total_sqft    365.9726
bath               1.0
price             18.5
bhk                  1
Name: 410, dtype: object

In [141]:
def perch_to_sqft(x):
    try:
        return int(float(x.split('P')[0])*272.25)
    except:
        return x

In [142]:
df4['total_sqft'] = df4['total_sqft'].apply(lambda x:perch_to_sqft(x))

In [143]:
df4

location       size  total_sqft  bath   price bhk
0      Electronic City Phase II      2 BHK  11366.6784   2.0   39.07   2
1              Chikka Tirupathi  4 Bedroom    27986.14   5.0  120.00   4
2                   Uttarahalli      3 BHK   15500.016   2.0   62.00   3
3            Lingadheeranahalli      3 BHK  16371.8919   3.0   95.00   3
4                      Kothanur      2 BHK    12916.68   2.0   51.00   2
...                         ...        ...         ...   ...     ...  ..
13315                Whitefield  5 Bedroom  37167.7467   4.0  231.00   5
13316             Richards Town      4 BHK    38750.04   5.0  400.00   4
13317     Raja Rajeshwari Nagar      2 BHK  12281.6099   2.0   60.00   2
13318           Padmanabhanagar      4 BHK  50471.9271   4.0  488.00   4
13319              Doddathoguru      1 BHK    5920.145   1.0   17.00   1

[13246 rows x 6 columns]

In [144]:
def Acres_to_sqft(x):
    try:
        return int(float(x.split('A')[0])*43560)
    except:
        return x
    


In [145]:
df4['total_sqft'] = df4['total_sqft'].apply(lambda x:Acres_to_sqft(x))

In [146]:
def convert(x):
    try:
        return int(float(x.split('-')[0]) + float(x.split('-')[1]))
    except:
        return x

In [147]:
df4['total_sqft'] =df4['total_sqft'].apply(lambda x:convert(x))

In [148]:
def Cents_to_sqft(x):
    try:
        return int(x.split('C')[0])*435.56 
    except:
        return x

In [149]:
df4['total_sqft'] =df4['total_sqft'].apply(lambda x:Cents_to_sqft(x))

In [150]:
def Guntha_to_sqft(x):
    try:
     return int(x.split('G')[0])*1089
    except:
       return x

In [151]:
df4['total_sqft'] =df4['total_sqft'].apply(lambda x:Guntha_to_sqft(x))

In [152]:
pd.to_numeric(df4['total_sqft'])

0        11366.6784
1        27986.1400
2        15500.0160
3        16371.8919
4        12916.6800
            ...    
13315    37167.7467
13316    38750.0400
13317    12281.6099
13318    50471.9271
13319     5920.1450
Name: total_sqft, Length: 13246, dtype: float64

In [153]:
df5 = df4.copy()

In [154]:
df5['price_per_sqft'] = df5['price']*100000 / df5['total_sqft']

In [155]:
df5.head(2)

location       size  total_sqft  bath   price bhk  \
0  Electronic City Phase II      2 BHK  11366.6784   2.0   39.07   2   
1          Chikka Tirupathi  4 Bedroom  27986.1400   5.0  120.00   4   

   price_per_sqft  
0      343.723985  
1      428.783676

In [156]:
len(df5.location.unique())

1304

In [157]:
df5.groupby(['location']).size().sort_values(ascending=False)

location
Whitefield             534
Sarjapur  Road         392
Electronic City        302
Kanakpura Road         266
Thanisandra            233
                      ... 
 Banaswadi               1
Kanakadasa Layout        1
Kanakapur main road      1
Kanakapura  Rod          1
whitefiled               1
Length: 1304, dtype: int64

In [158]:
location_stats = df5.groupby(['location']).size().sort_values(ascending=False)

In [159]:
location_less_than_10 = location_stats[location_stats<=10]

In [160]:
df5['location'] = df5['location'].apply(lambda x: 'other' if x in location_less_than_10 else x)

In [161]:
df5

location       size  total_sqft  bath   price bhk  \
0      Electronic City Phase II      2 BHK  11366.6784   2.0   39.07   2   
1              Chikka Tirupathi  4 Bedroom  27986.1400   5.0  120.00   4   
2                   Uttarahalli      3 BHK  15500.0160   2.0   62.00   3   
3            Lingadheeranahalli      3 BHK  16371.8919   3.0   95.00   3   
4                      Kothanur      2 BHK  12916.6800   2.0   51.00   2   
...                         ...        ...         ...   ...     ...  ..   
13315                Whitefield  5 Bedroom  37167.7467   4.0  231.00   5   
13316                     other      4 BHK  38750.0400   5.0  400.00   4   
13317     Raja Rajeshwari Nagar      2 BHK  12281.6099   2.0   60.00   2   
13318           Padmanabhanagar      4 BHK  50471.9271   4.0  488.00   4   
13319              Doddathoguru      1 BHK   5920.1450   1.0   17.00   1   

       price_per_sqft  
0          343.723985  
1          428.783676  
2          399.999587  
3          580.262810  
4          394.838302  
...               ...  
13315      621.506603  
13316     1032.256999  
13317      488.535302  
13318      966.874118  
13319      287.155129  

[13246 rows x 7 columns]

In [162]:
len(df5.location.unique())

242

on an average each bedroom occupies 300sqft or more , we will check that

In [163]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   location        13246 non-null  object 
 1   size            13246 non-null  object 
 2   total_sqft      13246 non-null  float64
 3   bath            13246 non-null  float64
 4   price           13246 non-null  float64
 5   bhk             13246 non-null  object 
 6   price_per_sqft  13246 non-null  float64
dtypes: float64(4), object(3)
memory usage: 1.3+ MB


In [164]:
df5['bhk'] = pd.to_numeric(df5['bhk'])

In [165]:
df5.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13246 entries, 0 to 13319
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   location        13246 non-null  object 
 1   size            13246 non-null  object 
 2   total_sqft      13246 non-null  float64
 3   bath            13246 non-null  float64
 4   price           13246 non-null  float64
 5   bhk             13246 non-null  int64  
 6   price_per_sqft  13246 non-null  float64
dtypes: float64(4), int64(1), object(2)
memory usage: 1.3+ MB


In [166]:
df5[df5['total_sqft']/df5['bhk']<300]

location       size  total_sqft  bath  price  bhk  \
349              other  3 Bedroom    118.4029   3.0   74.0    3   
1020             other      1 BHK    161.4585   1.0   30.0    1   
1122             other  5 Bedroom    258.3336   2.0  150.0    5   
4086    Sarjapur  Road  4 Bedroom     10.7639   4.0  120.0    4   
4972             other      7 BHK     53.8195   7.0  115.0    7   
5453             other  6 Bedroom   1306.6800   6.0   75.0    6   
5668   Judicial Layout  5 Bedroom    484.3755   6.0  400.0    5   
11558       Whitefield  4 Bedroom    645.8340   4.0  218.0    4   
12252            other      5 BHK    258.3336   5.0   75.0    5   

       price_per_sqft  
349      6.249847e+04  
1020     1.858063e+04  
1122     5.806446e+04  
4086     1.114838e+06  
4972     2.136772e+05  
5453     5.739737e+03  
5668     8.258056e+04  
11558    3.375480e+04  
12252    2.903223e+04

These are errors or outliers

In [167]:
df6 = df5[~(df5['total_sqft']/df5['bhk']<300)]

In [168]:
df5.shape


(13246, 7)

In [169]:
df6.info()

<class 'pandas.core.frame.DataFrame'>
Index: 13237 entries, 0 to 13319
Data columns (total 7 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   location        13237 non-null  object 
 1   size            13237 non-null  object 
 2   total_sqft      13237 non-null  float64
 3   bath            13237 non-null  float64
 4   price           13237 non-null  float64
 5   bhk             13237 non-null  int64  
 6   price_per_sqft  13237 non-null  float64
dtypes: float64(4), int64(1), object(2)
memory usage: 827.3+ KB


In [170]:

df6.shape


(13237, 7)

In [171]:
for key , subdf in df6.groupby('location'):
    print(key , subdf)

 Devarachikkanahalli                    location       size  total_sqft  bath  price  bhk  \
152     Devarachikkanahalli      3 BHK  13454.8750   2.0   44.0    3   
341     Devarachikkanahalli      2 BHK  13454.8750   2.0   40.0    2   
2474    Devarachikkanahalli  2 Bedroom  12916.6800   2.0   83.0    2   
5030    Devarachikkanahalli      2 BHK  12593.7630   2.0   40.0    2   
5536    Devarachikkanahalli  3 Bedroom  12916.6800   3.0  160.0    3   
5771    Devarachikkanahalli      3 BHK  15338.5575   2.0   65.0    3   
6591    Devarachikkanahalli      2 BHK  10193.4133   2.0   43.0    2   
7536    Devarachikkanahalli      2 BHK  12163.2070   2.0   36.0    2   
7767    Devarachikkanahalli      3 BHK  15252.4463   2.0   76.0    3   
8268    Devarachikkanahalli      2 BHK  13239.5970   2.0   58.0    2   
8954    Devarachikkanahalli      3 BHK  18352.4495   3.0   75.0    3   
9894    Devarachikkanahalli      2 BHK  12012.5124   2.0   47.0    2   
10582   Devarachikkanahalli      2 BHK  106

In [172]:
def remove_outliers(df):
    df_out = pd.DataFrame()
    for key , subdf in df.groupby('location'):
     m = np.mean(subdf.price_per_sqft)
     st = np.std( subdf.price_per_sqft) 
     reduced_df = subdf[(subdf.price_per_sqft>(m-st)) & (subdf.price_per_sqft<=(m+st))] 
     df_out = pd.concat([df_out , reduced_df] , ignore_index=True)
    return df_out

    

In [173]:
df7 = remove_outliers(df6)

In [174]:
df7.shape

(11742, 7)

usually if there's a 4bhk , it has 5-6 bathrooms and not more than that.

In [175]:
df7[df7['bath'] > df7['bhk']+2]

location       size   total_sqft  bath   price  bhk  \
1688      Chamrajpet  6 Bedroom   16145.8500   9.0   230.0    6   
1837   Chikkabanavar  4 Bedroom   26479.1940   7.0    80.0    4   
4266        KR Puram  8 Bedroom   12916.6800  12.0   110.0    8   
5985      Nagasandra  4 Bedroom   75347.3000   8.0   450.0    4   
6574    Rajaji Nagar      5 BHK   80729.2500   8.0  1700.0    5   
7774     Thanisandra      3 BHK   19439.6034   6.0   116.0    3   
9362           other  9 Bedroom   35520.8700  14.0   500.0    9   
9545           other  8 Bedroom   10656.2610  12.0   120.0    8   
9677           other      6 BHK  122041.0982   9.0  1000.0    6   
11079          other  6 Bedroom   12916.6800   9.0   122.0    6   

       price_per_sqft  
1688      1424.514659  
1837       302.124000  
4266       851.612024  
5985       597.234407  
6574      2105.804278  
7774       596.719993  
9362      1407.623180  
9545      1126.098544  
9677       819.396101  
11079      944.515154

In [176]:
df8 = df7[df7['bath'] < df7['bhk']+2]

In [177]:
df8.shape

(11621, 7)

In [178]:
df8.drop(['size' , 'price_per_sqft'] , axis=1)

location  total_sqft  bath   price  bhk
0       Devarachikkanahalli  13454.8750   2.0   44.00    3
1       Devarachikkanahalli  13454.8750   2.0   40.00    2
2       Devarachikkanahalli  12916.6800   2.0   83.00    2
3       Devarachikkanahalli  12593.7630   2.0   40.00    2
4       Devarachikkanahalli  15338.5575   2.0   65.00    3
...                     ...         ...   ...     ...  ...
11737                 other  14563.5567   2.0  110.00    2
11738                 other   8740.2868   1.0   26.00    1
11739                 other  15500.0160   2.0   63.93    3
11740                 other  11571.1925   2.0   48.00    2
11741                 other  38750.0400   5.0  400.00    4

[11621 rows x 5 columns]

In [179]:
dummies = pd.get_dummies(df8['location'] , dtype=int , drop_first=True)

In [180]:
dummies.shape

(11621, 241)

In [181]:
dummies

1st Block Jayanagar  1st Phase JP Nagar  2nd Phase Judicial Layout  \
0                        0                   0                          0   
1                        0                   0                          0   
2                        0                   0                          0   
3                        0                   0                          0   
4                        0                   0                          0   
...                    ...                 ...                        ...   
11737                    0                   0                          0   
11738                    0                   0                          0   
11739                    0                   0                          0   
11740                    0                   0                          0   
11741                    0                   0                          0   

       2nd Stage Nagarbhavi  5th Block Hbr Layout  5th Phase JP Nagar  \
0                         0                     0                   0   
1                         0                     0                   0   
2                         0                     0                   0   
3                         0                     0                   0   
4                         0                     0                   0   
...                     ...                   ...                 ...   
11737                     0                     0                   0   
11738                     0                     0                   0   
11739                     0                     0                   0   
11740                     0                     0                   0   
11741                     0                     0                   0   

       6th Phase JP Nagar  7th Phase JP Nagar  8th Phase JP Nagar  \
0                       0                   0                   0   
1                       0                   0                   0   
2                       0                   0                   0   
3                       0                   0                   0   
4                       0                   0                   0   
...                   ...                 ...                 ...   
11737                   0                   0                   0   
11738                   0                   0                   0   
11739                   0                   0                   0   
11740                   0                   0                   0   
11741                   0                   0                   0   

       9th Phase JP Nagar  ...  Vishveshwarya Layout  Vishwapriya Layout  \
0                       0  ...                     0                   0   
1                       0  ...                     0                   0   
2                       0  ...                     0                   0   
3                       0  ...                     0                   0   
4                       0  ...                     0                   0   
...                   ...  ...                   ...                 ...   
11737                   0  ...                     0                   0   
11738                   0  ...                     0                   0   
11739                   0  ...                     0                   0   
11740                   0  ...                     0                   0   
11741                   0  ...                     0                   0   

       Vittasandra  Whitefield  Yelachenahalli  Yelahanka  Yelahanka New Town  \
0                0           0               0          0                   0   
1                0           0               0          0                   0   
2                0           0               0          0                   0   
3                0           0               0          0                   0   
4                0           0               0          0 

In [182]:
df9 = pd.concat([df8 , dummies] , axis=1)

In [183]:
df9.head(5)

location       size  total_sqft  bath  price  bhk  \
0   Devarachikkanahalli      3 BHK  13454.8750   2.0   44.0    3   
1   Devarachikkanahalli      2 BHK  13454.8750   2.0   40.0    2   
2   Devarachikkanahalli  2 Bedroom  12916.6800   2.0   83.0    2   
3   Devarachikkanahalli      2 BHK  12593.7630   2.0   40.0    2   
4   Devarachikkanahalli      3 BHK  15338.5575   2.0   65.0    3   

   price_per_sqft  1st Block Jayanagar  1st Phase JP Nagar  \
0      327.019017                    0                   0   
1      297.290016                    0                   0   
2      642.579982                    0                   0   
3      317.617538                    0                   0   
4      423.768663                    0                   0   

   2nd Phase Judicial Layout  ...  Vishveshwarya Layout  Vishwapriya Layout  \
0                          0  ...                     0                   0   
1                          0  ...                     0                   0   
2                          0  ...                     0                   0   
3                          0  ...                     0                   0   
4                          0  ...                     0                   0   

   Vittasandra  Whitefield  Yelachenahalli  Yelahanka  Yelahanka New Town  \
0            0           0               0          0                   0   
1            0           0               0          0                   0   
2            0           0               0          0                   0   
3            0           0               0          0                   0   
4            0           0               0          0                   0   

   Yelenahalli  Yeshwanthpur  other  
0            0             0      0  
1            0             0      0  
2            0             0      0  
3            0             0      0  
4            0             0      0  

[5 rows x 248 columns]

In [184]:
df10 = df9.drop(['location' , 'price_per_sqft' , 'size'] , axis=1)

In [185]:
df10.head(10)

total_sqft  bath  price  bhk  1st Block Jayanagar  1st Phase JP Nagar  \
0  13454.8750   2.0   44.0    3                    0                   0   
1  13454.8750   2.0   40.0    2                    0                   0   
2  12916.6800   2.0   83.0    2                    0                   0   
3  12593.7630   2.0   40.0    2                    0                   0   
4  15338.5575   2.0   65.0    3                    0                   0   
5  10193.4133   2.0   43.0    2                    0                   0   
6  12163.2070   2.0   36.0    2                    0                   0   
7  15252.4463   2.0   76.0    3                    0                   0   
8  13239.5970   2.0   58.0    2                    0                   0   
9  18352.4495   3.0   75.0    3                    0                   0   

   2nd Phase Judicial Layout  2nd Stage Nagarbhavi  5th Block Hbr Layout  \
0                          0                     0                     0   
1                          0                     0                     0   
2                          0                     0                     0   
3                          0                     0                     0   
4                          0                     0                     0   
5                          0                     0                     0   
6                          0                     0                     0   
7                          0                     0                     0   
8                          0                     0                     0   
9                          0                     0                     0   

   5th Phase JP Nagar  ...  Vishveshwarya Layout  Vishwapriya Layout  \
0                   0  ...                     0                   0   
1                   0  ...                     0                   0   
2                   0  ...                     0                   0   
3                   0  ...                     0                   0   
4                   0  ...                     0                   0   
5                   0  ...                     0                   0   
6                   0  ...                     0                   0   
7                   0  ...                     0                   0   
8                   0  ...                     0                   0   
9                   0  ...                     0                   0   

   Vittasandra  Whitefield  Yelachenahalli  Yelahanka  Yelahanka New Town  \
0            0           0               0          0                   0   
1            0           0               0          0                   0   
2            0           0               0          0                   0   
3            0           0               0          0                   0   
4            0           0               0          0                   0   
5            0           0               0          0                   0   
6            0           0               0          0                   0   
7            0           0               0          0                   0   
8            0           0               0          0                   0   
9            0           0               0          0                   0   

   Yelenahalli  Yeshwanthpur  other  
0            0             0      0  
1            0             0      0  
2            0             0      0  
3            0             0      0  
4            0             0      0  
5            0             0      0  
6            0             0      0  
7            0             0      0  
8            0             0      0  
9            0             0      0  

[10 rows x 245 columns]

In [186]:
x = df10.drop(['price'] , axis=1)

In [187]:
x.head(2)

total_sqft  bath  bhk  1st Block Jayanagar  1st Phase JP Nagar  \
0   13454.875   2.0    3                    0                   0   
1   13454.875   2.0    2                    0                   0   

   2nd Phase Judicial Layout  2nd Stage Nagarbhavi  5th Block Hbr Layout  \
0                          0                     0                     0   
1                          0                     0                     0   

   5th Phase JP Nagar  6th Phase JP Nagar  ...  Vishveshwarya Layout  \
0                   0                   0  ...                     0   
1                   0                   0  ...                     0   

   Vishwapriya Layout  Vittasandra  Whitefield  Yelachenahalli  Yelahanka  \
0                   0            0           0               0          0   
1                   0            0           0               0          0   

   Yelahanka New Town  Yelenahalli  Yeshwanthpur  other  
0                   0            0             0      0  
1                   0            0             0      0  

[2 rows x 244 columns]

In [188]:
y = df10['price']

In [189]:
y

0         44.00
1         40.00
2         83.00
3         40.00
4         65.00
          ...  
11737    110.00
11738     26.00
11739     63.93
11740     48.00
11741    400.00
Name: price, Length: 11621, dtype: float64

In [190]:
from sklearn.model_selection import train_test_split

In [191]:
x_train , x_test , y_train , y_test = train_test_split(x , y , test_size=0.2 , random_state=5)

In [192]:
from sklearn.model_selection import cross_val_score 
from sklearn.model_selection import ShuffleSplit

In [193]:
cv = ShuffleSplit(n_splits=5 , test_size=0.2 , random_state=5)

In [194]:
from sklearn.linear_model import LinearRegression
from sklearn.linear_model import Lasso
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
cross_val_score(LinearRegression() ,x , y , cv=cv)

array([-0.00110495,  0.22227321,  0.17591709,  0.67901368,  0.65501385])

In [195]:
cross_val_score(Lasso() ,x , y , cv=cv)

array([-0.15604911,  0.08567394,  0.05353909,  0.60755203,  0.57163739])

In [196]:
cross_val_score(DecisionTreeRegressor() ,x , y , cv=cv)

array([0.55208811, 0.53839815, 0.48107686, 0.66155393, 0.72564526])

In [197]:
cross_val_score(RandomForestRegressor() ,x , y , cv=cv)

array([0.72833181, 0.76594149, 0.74497009, 0.78689948, 0.79625988])

In [198]:
from sklearn.neighbors import KNeighborsRegressor

In [199]:
cross_val_score(KNeighborsRegressor() ,x , y , cv=cv)

array([0.51415448, 0.65149196, 0.64262663, 0.63023873, 0.63083205])

random forest regressor gives overall better results

In [200]:
# cross_val_score(RandomForestRegressor(n_estimators=300) ,x , y , cv=cv)

n_estimators are hardly making a difference

In [201]:
model = RandomForestRegressor()

In [202]:
model.fit(x_train, y_train)

RandomForestRegressor()

In [203]:
model.score(x_test , y_test)

0.7189007494615134

In [204]:
np.where(x.columns == 'Yelahanka')[0][0]

239

In [205]:
x.head()

total_sqft  bath  bhk  1st Block Jayanagar  1st Phase JP Nagar  \
0  13454.8750   2.0    3                    0                   0   
1  13454.8750   2.0    2                    0                   0   
2  12916.6800   2.0    2                    0                   0   
3  12593.7630   2.0    2                    0                   0   
4  15338.5575   2.0    3                    0                   0   

   2nd Phase Judicial Layout  2nd Stage Nagarbhavi  5th Block Hbr Layout  \
0                          0                     0                     0   
1                          0                     0                     0   
2                          0                     0                     0   
3                          0                     0                     0   
4                          0                     0                     0   

   5th Phase JP Nagar  6th Phase JP Nagar  ...  Vishveshwarya Layout  \
0                   0                   0  ...                     0   
1                   0                   0  ...                     0   
2                   0                   0  ...                     0   
3                   0                   0  ...                     0   
4                   0                   0  ...                     0   

   Vishwapriya Layout  Vittasandra  Whitefield  Yelachenahalli  Yelahanka  \
0                   0            0           0               0          0   
1                   0            0           0               0          0   
2                   0            0           0               0          0   
3                   0            0           0               0          0   
4                   0            0           0               0          0   

   Yelahanka New Town  Yelenahalli  Yeshwanthpur  other  
0                   0            0             0      0  
1                   0            0             0      0  
2                   0            0             0      0  
3                   0            0             0      0  
4                   0            0             0      0  

[5 rows x 244 columns]

In [206]:
np.zeros(len(x.columns))

array([0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0., 0.,
       0., 0., 0., 0., 0.

In [207]:
np.zeros(5)

array([0., 0., 0., 0., 0.])

In [208]:
x.head()

total_sqft  bath  bhk  1st Block Jayanagar  1st Phase JP Nagar  \
0  13454.8750   2.0    3                    0                   0   
1  13454.8750   2.0    2                    0                   0   
2  12916.6800   2.0    2                    0                   0   
3  12593.7630   2.0    2                    0                   0   
4  15338.5575   2.0    3                    0                   0   

   2nd Phase Judicial Layout  2nd Stage Nagarbhavi  5th Block Hbr Layout  \
0                          0                     0                     0   
1                          0                     0                     0   
2                          0                     0                     0   
3                          0                     0                     0   
4                          0                     0                     0   

   5th Phase JP Nagar  6th Phase JP Nagar  ...  Vishveshwarya Layout  \
0                   0                   0  ...                     0   
1                   0                   0  ...                     0   
2                   0                   0  ...                     0   
3                   0                   0  ...                     0   
4                   0                   0  ...                     0   

   Vishwapriya Layout  Vittasandra  Whitefield  Yelachenahalli  Yelahanka  \
0                   0            0           0               0          0   
1                   0            0           0               0          0   
2                   0            0           0               0          0   
3                   0            0           0               0          0   
4                   0            0           0               0          0   

   Yelahanka New Town  Yelenahalli  Yeshwanthpur  other  
0                   0            0             0      0  
1                   0            0             0      0  
2                   0            0             0      0  
3                   0            0             0      0  
4                   0            0             0      0  

[5 rows x 244 columns]

In [209]:
x.columns

Index(['total_sqft', 'bath', 'bhk', '1st Block Jayanagar',
       '1st Phase JP Nagar', '2nd Phase Judicial Layout',
       '2nd Stage Nagarbhavi', '5th Block Hbr Layout', '5th Phase JP Nagar',
       '6th Phase JP Nagar',
       ...
       'Vishveshwarya Layout', 'Vishwapriya Layout', 'Vittasandra',
       'Whitefield', 'Yelachenahalli', 'Yelahanka', 'Yelahanka New Town',
       'Yelenahalli', 'Yeshwanthpur', 'other'],
      dtype='object', length=244)

In [214]:
def predict_price(location , sqft , bath , bhk):
    loc_index = np.where(x.columns == location )[0][0]
    a=np.zeros( len(x.columns))
    a[0] = sqft
    a[1] = bath
    a[2] = bhk
    if loc_index >= 0:
        a[loc_index] = 1

    return model.predict([a])[0]

    

In [229]:
predict_price('Whitefield',1000,3,3)

c:\Users\SWARAJ\AppData\Local\Programs\Python\Python312\Lib\site-packages\sklearn\base.py:493: UserWarning: X does not have valid feature names, but RandomForestRegressor was fitted with feature names
  warnings.warn(


27.44

In [212]:
df6['price_per_sqft'].loc[7]

1689.1478164808466

In [213]:
remove_outliers(df6['price_per_sqft'])

KeyError: 'location'

In [ ]:
df5['location'].apply(lambda x: x.strip())
# to remove extra spaces before and after the words

0        Electronic City Phase II
1                Chikka Tirupathi
2                     Uttarahalli
3              Lingadheeranahalli
4                        Kothanur
                   ...           
13315                  Whitefield
13316                       other
13317       Raja Rajeshwari Nagar
13318             Padmanabhanagar
13319                Doddathoguru
Name: location, Length: 13246, dtype: object